### ISSUE TITLE
Updating to Context Search 4.0.0 removes all custom search engines and loads default engines at start

### ISSUE THREAD

**Issue by mudbone79** [odebroqueville/contextSearch#82]
I am on Firefox 79 (64bit).  Every time I startup Firefox Context Search returns to the default search engines and I have to reload my custom sear engine list. I think this just started with the last Firefox update but not 100% sure.  I did notice another user mentioning this issue in the reviews.

**Comment by odebroqueville** [comment]
@mudbone79 Would you please tell us which version of the extension you are using and for which OS?

**Comment by mudbone79** [comment]
The extension version is 3.72.  I am running Windows 10 Home 64bit v2004 build 19041.450

**Comment by ghmv** [comment]
Well, for me it just happened when it upgraded to 4.0. All the search engines are gone. I'm pissed. I have not backuped it – who would think that the extension would just delete it. What to do? (Firefox 80b8, but it worked recently on Firefox 80b8 so that's not the problem.)

**Comment by ObJuanK** [comment]
I am using Firefox v79 and Addon version 4.0 and this happens daily. Every time i close my browser and open it again, all my search engines are gone. Thankfully I have backed them up, but I have to restore them every time

**Comment by ghmv** [comment]
After repairing this problem, it would help adding at least a function for copying search engines from the browser (or even directly using them) because nearly nobody will repair all the deleted search engines. I'm going to abandon the extension because it was some work to set it and it's clearly undependable now.

**Comment by MikhailAristov** [comment]
Just wanted to chime in that the same issue has just happened to me, after the automatic upgrade to add-on version 4.0 on Firefox 79.0 x64. I was lucky enough to have had a backup ready to re-import the engines, but if this thing happens every day, as this thread suggest, I will have to look for alternatives until it's fixed.

**Comment by odebroqueville** [comment]
@cjscheller CS 3.72 stored search engines along with the extension's options in storage sync. CS 4.0.0 still stores the options in storage sync but it stores the search engines in local storage. So, I assume what is happening is that b/c the CS 4.0.0 doesn't find any search engines in local storage, it loads the default set of search engines. Why it keeps reloading the default set of search engines after restarting the browser even after users have imported their custom set of search engines puzzles me!
@MikhailAristov Can you pls do the following and tell us what happens:
1. Remove Context Search 4.0.0 from your browser
2. Restart your browser
3. Re-install CS 4.0.0
4. Open a new tab and type 'cs .' (without the quotes) to open the extension's prefs page
5. Import your custom set of search engines
6. Re-start your browser
7. Open a new tab and type 'cs .' again to verify that your custom set has been saved
8. Open a new tab and visit any web page (e.g. bbc.com) and test that your search engines appear in the context menu when right-clicking on a text selection
Pls report your findings.

**Comment by odebroqueville** [comment]
@ghmv I'm sorry to hear you didn't backup your list of search engines, but I think that we may still recover your list and make a backup.
The first thing you need to do is identify your browser profile. For this, open your browser and visit Help > Troubleshooting Information. This will open a new tab and somewhere in the left margin you will see 'Profile Folder' and a button to open the folder in the Finder.
In this folder you should see a file named storage-sync.sqlite. Yous should copy this file to a safe location somewhere on your disk b/c I believe it still contains your list of search engines. You can verify if this is the case by visiting the App Store and donwloading an app that will allow you to view the contents of the sqlite file. (In the App Store, perform a search with 'sqlite' to list apps that can do this. Some are free, others not!)
Next, what you need to do is remove any Context Search version currently installed, then head over to https://addons.mozilla.org/en-US/firefox/addon/contextual-search/versions/
and re-install version 3.72.
Then you need to visit the extensions preferences page. If your extensions aren't listed, then head over to your browser profile folder and replace the storage-sync.sqlite folder with the one you copied to a safe place previously.
Always MAKE SURE TO KEEP THE ORIGINAL storage-sync.sqlite FILE AS A COPY IN A SAFE PLACE ON YOUR DISK.
If everything went according to plans, when you now visit the extensions prefs page, you should now see your list of custom search engines. Make a backup now by exporting your list and keep it in a safe spot.
I hope this will work for you.

**Comment by kuality** [comment]
I have the same problem on two computers where I use ContextSearch: at some point today the search engines disappeared. FF79.0 ContextSearch 4.0.0 updated Aug 25.
I tried your suggestion to copy the storage-sync.sqlite file, but I don't know how I would re-install 3.72, since the add-on store doesn't seem to give version options. 
You guys have built a great tool, I am really missing the functionality, I hope you manage a fix soon!

**Comment by MikhailAristov** [comment]
> @MikhailAristov Can you pls do the following and tell us what happens:
> 
>     1. Remove Context Search 4.0.0 from your browser
> 
>     2. Restart your browser
> 
>     3. Re-install CS 4.0.0
> 
>     4. Open a new tab and type 'cs .' (without the quotes) to open the extension's prefs page
> 
>     5. Import your custom set of search engines
> 
>     6. Re-start your browser
> 
>     7. Open a new tab and type 'cs .' again to verify that your custom set has been saved
> 
>     8. Open a new tab and visit any web page (e.g. bbc.com) and test that your search engines appear in the context menu when right-clicking on a text selection
>        Pls report your findings.

Sure, I did that just now. First thing I noticed was the FF updated to v80.0 after restart, not sure if that affects anything. Upon reinstallation, CS 4.0 started out with the default set of search engines, but after browser restart, it has kept my imported custom set, and it shows up properly in the context menu. Problem solved, I guess? :D

**Comment by ObJuanK** [comment]
> 
> 
> > @MikhailAristov Can you pls do the following and tell us what happens:
> > ```
> > 1. Remove Context Search 4.0.0 from your browser
> > 
> > 2. Restart your browser
> > 
> > 3. Re-install CS 4.0.0
> > 
> > 4. Open a new tab and type 'cs .' (without the quotes) to open the extension's prefs page
> > 
> > 5. Import your custom set of search engines
> > 
> > 6. Re-start your browser
> > 
> > 7. Open a new tab and type 'cs .' again to verify that your custom set has been saved
> > 
> > 8. Open a new tab and visit any web page (e.g. bbc.com) and test that your search engines appear in the context menu when right-clicking on a text selection
> >    Pls report your findings.
> > ```
> 
> Sure, I did that just now. First thing I noticed was the FF updated to v80.0 after restart, not sure if that affects anything. Upon reinstallation, CS 4.0 started out with the default set of search engines, but after browser restart, it has kept my imported custom set, and it shows up properly in the context menu. Problem solved, I guess? :D

Can confirm that I did the same steps and got the same result as @MikhailAristov 
Upgraded to V80 of FF after restart and search engines remain after subsequent restarts of the browser

**Comment by odebroqueville** [comment]
@kuality you may access older versions, including version 3.72, here: https://addons.mozilla.org/en-US/firefox/addon/contextual-search/versions/

**Comment by kuality** [comment]
> 
> 
> @kuality you may access older versions, including version 3.72, here: https://addons.mozilla.org/en-US/firefox/addon/contextual-search/versions/
I was able to install the old version and find my search preferences intact, but I can't export them: when I click the export button nothing at all happens. The import button starts a file dialog - that doesn't help, just for context.

**Comment by odebroqueville** [comment]
@kuality have you checked your downloads folder for a file named searchEngines.json ?

**Comment by kuality** [comment]
> @kuality have you checked your downloads folder for a file named searchEngines.json ?

I didn't then, because there was no indication at all that something was being downloaded. I did now, and there is no searchEngines.json file in my downloads folder.

**Comment by mudbone79** [comment]
Uninstalling and re-installing per the instructions given worked for me.

**Comment by odebroqueville** [comment]
@kuality would you mind sending us a copy of your storage-sync.sqlite file so we can try and recover your list of search engines?

**Comment by kuality** [comment]
> @kuality would you mind sending us a copy of your storage-sync.sqlite file so we can try and recover your list of search engines?
I would rather not send that file. I have re-created the list, I was just making sure to post what didn't work in the interest of bug reporting to make sure all of those people who don't report also get the fix.

**Comment by ghmv** [comment]
Thanks Olivier. All search engines were in storage-sync.sqlite. 
But after upgrading to 4.0 from 3.7.2, it again resets to default search engines. 
When reloaded again from the backup, then it remains in the extension settings when browser is reloaded.
(Firefox Nightly)

**Comment by Ainson** [comment]
Same happened to me. Updated Context Search and all my searches were wiped out, replaced by the default. (I got this message in a popup: _Default list of search engines has been loaded_)

The setting _"Force reload of default search engines list_" is unchecked.

Before the default searches would add itself to my list, but this time it replaced them.

Luckily I had a recent backup so no big deal.

**Comment by giethmon** [comment]
Hi, same problem on Ff 80.0.1/Win 10 with **active Ff-sync**. Could the error be caused by https://blog.mozilla.org/addons/2020/07/09/changes-to-storage-sync-in-firefox-79/ ? As far as I can see all (old) settings go to storage-sync.sqlite. 
There are three new files in profile: storage-sync-v2.sqlite, storage-sync-v2.sqlite-shm and storage-sync-v2.sqlite-wal.

**Comment by odebroqueville** [comment]
@giethmon Yeah, we first thought that the bug was caused by the changes introduced in Firefox 79, but we were unable to reproduce the bug using a simplified extension. We filed a bug report on Bugzilla, but we're pretty much left having to test and resolve things on our side. For those who've already updated to Context Search 4.0.0 and ffx 79, the only solution we have is to re-install Context Search and import your custom list of search engines (hoping everyone has a backup!).
https://bugzilla.mozilla.org/show_bug.cgi?id=1661325
